compiler/x/codegen/OMRAheadOfTimeCompile.cpp

```diff
@@ -0,0 +1,142 @@
+      }
+    if (self()->getSizeOfAOTRelocations() != 0)
+      {
+      uint8_t *relocationDataCursor = self()->setRelocationData(fe->allocateRelocationData(self()->comp(), self()->getSizeOfAOTRelocations() + sizeof(uintptr_t )));
+      uint8_t * copy = relocationDataCursor;
+      // set up the size for the region
+      TR::RelocationRecordBinaryTemplate* groups = reinterpret_cast<
+      TR::RelocationRecordBinaryTemplate*> (relocationDataCursor);
+      TR::RelocationRuntime *reloRuntime =self()->reloRuntime();
+      TR::RelocationTarget *reloTarget = reloRuntime->reloTarget();
+      OMR::RelocationRecordGroup reloGroup(groups);
+
+      reloGroup.setSize(reloTarget,self()->getSizeOfAOTRelocations()+ OMR::AheadOfTimeCompile::SIZEPOINTER);
+            self()->addToSizeOfAOTRelocations(OMR::AheadOfTimeCompile::SIZEPOINTER);
+      relocationDataCursor +=OMR::AheadOfTimeCompile::SIZEPOINTER;
+      TR::IteratedExternalRelocation *s;
+      for (s = self()->getAOTRelocationTargets().getFirst();
+           s != NULL;
+           s = s->getNext())
+         {
+         s->setRelocationData(relocationDataCursor);
+         s->initializeRelocation(cg);
+         relocationDataCursor += s->getSizeOfRelocationData();
+         }
+      }
+}
+
+uint8_t* OMR::X86::AheadOfTimeCompile::initializeAOTRelocationHeader(TR::IteratedExternalRelocation *relocation)
+   {
+   TR::Compilation* comp = self()->comp();
+   TR::CodeGenerator* cg = comp->cg();
+   TR::RelocationRuntime *reloRuntime =self()->reloRuntime();
+   TR::RelocationTarget *reloTarget = reloRuntime->reloTarget();
+   uint8_t* cursor;
+   TR::RelocationRecordBinaryTemplate* binaryTemplate = reinterpret_cast<TR::RelocationRecordBinaryTemplate *>(relocation->getRelocationData());
+   uint16_t sizeOfRelocationData = relocation->getSizeOfRelocationData();
+   binaryTemplate->setSize(reloTarget,sizeOfRelocationData);
+   uint8_t targetKind = relocation->getTargetKind();
+   binaryTemplate->setType(reloTarget,targetKind);
+   uint8_t  wideOffsets = relocation->needsWideOffsets() ? RELOCATION_TYPE_WIDE_OFFSET : 0;
+   binaryTemplate->setFlags(reloTarget,wideOffsets);
+
+   // This has to be created after the kind has been written into the header
+   TR::RelocationRecord storage;
+   TR::RelocationRecord *reloRecord = TR::RelocationRecord::create(&storage, 
+                                       reloRuntime->self(), reloTarget,
+                                       reinterpret_cast<TR::RelocationRecordBinaryTemplate *>
+                                       (relocation->getRelocationData()));
+   switch (targetKind)
+      {
+         default:
+         // initializeCommonOMRAOTRelocationHeader is currently in the process
+         // of becoming the canonical place to initialize the platform agnostic
+         // relocation headers; new relocation records' header should be
+         // initialized here.
+         cursor = self()->initializeCommonOMRAOTRelocationHeader(relocation, reloRecord);
```

Comment:
isn't this function marked as TR_UNIMPLEMENTED in the base class?

"OMRAOT" seems redundant in this function name.

Reply:
Hello @mstoodle, thank you for comment. This name is used to avoid overriding this method with openJ9. The problem was in when I was building with autoconf and cmake, one of those build systems was complaining. Maybe an alternative name should be used, can you suggest one?
Regarding the TR_UNIMPLEMENTED: the initializeCommonOMRAOTRelocationHeader was a desired way to create architecture independent relocation records, however I did not have any implementation for initializeCommonOMRAOTRelocationHeader since we did not identify relocation records that belong to OMR and are not project specific. How should I address the problem?

Reply:
I guess it's ok, but probably makes it harder to test all this code in OMR itself :( .

Reply:
We have a test compiler extension we worked on over summer, this is a next step in #5358